### ISSUE TITLE
What happened to variants in PR #120?

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#540]
The sample packages.yaml file given in the new concretization docs does not parse:

```
 +    packages:
+      dyninst:
+        compiler: [gcc@4.9]
+        variants: +debug
+      gperftools:
+        version: [2.2, 2.4, 2.3]
+      all:
+        compiler: [gcc@4.4.7, gcc@4.6:, intel, clang, pgi]
+        providers:
+          mpi: [mvapich, mpich, openmpi]
```

The yaml validation complains on any "variants:" line.  The spec for packages.yaml (in config.py) confirms that the 'variants:' line is not (currently) legal in packages.yaml:

Was it the intention to include the 'variants' feature in PR #120?  I'm hoping so, since it will be really useful for me.

```
    'packages': {
        '$schema': 'http://json-schema.org/schema#',
        'title': 'Spack package configuration file schema',
        'type': 'object',
        'additionalProperties': False,
        'patternProperties': {
            r'packages:?': {
                'type': 'object',
                'default': {},
                'additionalProperties': False,
                'patternProperties': {
                    r'\w[\w-]*': { # package name
                        'type': 'object',
                        'default': {},
                        'additionalProperties': False,
                        'properties': {
                            'version': {
                                'type' : 'array',
                                'default' : [],
                                'items' : { 'anyOf' : [ { 'type' : 'string' },
                                                        { 'type' : 'number'}]}}, #version strings
                            'compiler': {
                                'type' : 'array',
                                'default' : [],
                                'items' : { 'type' : 'string' } }, #compiler specs
                            'buildable': {
                                'type':  'boolean',
                                'default': True,
                             },
                            'providers': {
                                'type':  'object',
                                'default': {},
                                'additionalProperties': False,
                                'patternProperties': {
                                    r'\w[\w-]*': {
                                        'type' : 'array',
                                        'default' : [],
                                        'items' : { 'type' : 'string' },},},},
                            'paths': {
                                'type' : 'object',
                                'default' : {},
                            }
                        },},},},},}
```

**Comment by mplegendre** [comment]
That's a documentation bug. Variants weren't intended to be supported in PR #120.  I'll remove that line. 
Variants certainly are the next sensible step in this work.  I'd like to get to them, though I don't know exactly when that will happen.

**Comment by citibeth** [comment]
Are there any objections if I take a crack at it?  Any suggestions on where to start in the code?

**Comment by mplegendre** [comment]
Not objections at all, and I'd be happy to help/review/test as needed.

A few things will be needed:

- In spack/config.py we'll need to expand the allowed package.yaml format to be able to specify variants and values for those variants.

- In concretize_variants() in spack/concretize.py we currently iterate over each variant in a package, and if it's not explicitly specified set the variant's value to its default.  We'll need to change that so that if it's not explicitly specified, then try the values set in the package.yaml config file, then if those aren't set go to the default values.

This is different than how we configure compilers/versions/providers, where we provide an ordered list of which options we most prefer to least prefer.  I don't think it makes as much sense to specify variant values in that way.

- We'll need tests/documentation.

**Comment by citibeth** [comment]
Thank you!

**Comment by mamelara** [comment]
Not sure what the status is on this but having static be the default when building packages on Cori would be very nice.

**Comment by adamjstewart** [comment]
This issue is resolved by #1208 and can be closed now.
